src/file_output.rs

```diff
@@ -0,0 +1,39 @@
+use crate::{Sample, Source};
+use hound::{SampleFormat, WavSpec};
+use std::path;
+
+/// This procedure saves Source's output into a wav file. The output samples format is 32-bit float.
+/// This function is intended primarily for testing and diagnostics. It can be used to see
+/// the output without opening output stream to a real audio device.
+pub fn output_to_wav<S: Sample, P: AsRef<path::Path>, Src: Source<Item = S>>(
+    source: &mut Src,
+    wav_file: &P,
```

Comment:
`wav_file: impl AsRef<Path>`, you can do the same for source, then you only need the generic `parameter S` for sample. Its more readable since you do not need to check what &P is

Reply:
Do you mean
```rust
pub fn output_to_wav<S: Sample>(
    source: &mut dyn Source<Item=S>,
    wav_file: &dyn AsRef<path::Path>,
) 
```
?

Reply:
No normal static dispatch but using a different notation:
```rust
pub fn output_to_wav<S: Sample>(
    source: &mut impl Source<Item=S>,
    wav_file: &impl AsRef<path::Path>,
) 
```

Reply:
actually I do not think you need the & in front of `impl AsRef<path::Path>`, since if AsRef is implemented for `thing` its also implemented for `&thing`

Reply:
// It was a misprint, it should have been `&dyn AsRef`

`AsRef` is a trait so it is unsized... If we want it to keep generic it should be either `dyn` or a template parameter? Can you give an example of what you have in mind?

Reply:
@dvdsk Updated.